File: src/main/java/com/rescue/mesh/model/MeshPacket.java
package com.rescue.mesh.model;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.annotations.SerializedName;
import com.rescue.mesh.util.ChecksumUtil;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

/**
 * Cấu trúc gói tin chuẩn canonical v1.0 của hệ thống Emergency Mesh Rescue.
 *
 * Chuẩn hóa schema JSON snake_case theo tài liệu docs/protocol/mesh-packet-v1.md.
 * Hỗ trợ đọc cả alias cũ (SOS_DATA, DISPATCH_CMD, camelCase) để đảm bảo tương thích ngược.
 *
 * QUAN TRỌNG: Checksum SHA-256 là mã băm kiểm tra tính toàn vẹn (Data Integrity),
 * không phải cơ chế xác thực danh tính (Authentication).
 */
public class MeshPacket {

    // ===== PHIÊN BẢN GIAO THỨC =====
    public static final String PROTOCOL_VERSION_1 = "1.0";

    // ===== HẰNG SỐ LOẠI GÓI TIN CANONICAL V1 =====
    public static final String TYPE_SOS_BROADCAST    = "SOS_BROADCAST";
    public static final String TYPE_DISPATCH_COMMAND  = "DISPATCH_COMMAND";
    public static final String TYPE_ROUTE_DISCOVERY  = "ROUTE_DISCOVERY";
    public static final String TYPE_ACK              = "ACK";
    public static final String TYPE_HEARTBEAT        = "HEARTBEAT";

    // ===== ALIAS CŨ CHO TƯƠNG THÍCH NGƯỢC (DEPRECATED) =====
    @Deprecated
    public static final String TYPE_SOS_DATA         = "SOS_DATA";
    @Deprecated
    public static final String TYPE_DISPATCH_CMD     = "DISPATCH_CMD";

    // ===== HẰNG SỐ MỨC ĐỘ NGUY HIỂM =====
    public static final String SEVERITY_CRITICAL     = "CRITICAL";
    public static final String SEVERITY_HIGH         = "HIGH";
    public static final String SEVERITY_MEDIUM       = "MEDIUM";

    // ===== HẰNG SỐ LOẠI TÌNH HUỐNG =====
    public static final String ALERT_MEDICAL         = "MEDICAL";
    public static final String ALERT_FLOOD           = "FLOOD_TRAPPED";
    public static final String ALERT_LANDSLIDE       = "LANDSLIDE";

    // ===== HẰNG SỐ NODE ID =====
    public static final String NODE_BASE_STATION     = "BASE_STATION";
    public static final String NODE_A_VICTIM         = "NODE_A_VICTIM";
    public static final String NODE_B_RELAY          = "NODE_B_RELAY";
    public static final String NODE_BROADCAST        = "BROADCAST";

    // ===== TTL MẶC ĐỊNH =====
    public static final int DEFAULT_TTL              = 5;

    // ===== CANONICAL FIELDS =====

    @SerializedName(value = "packet_id", alternate = {"packetId"})
    private String packetId;

    @SerializedName(value = "protocol_version", alternate = {"protocolVersion"})
    private String protocolVersion = PROTOCOL_VERSION_1;

    @SerializedName(value = "packet_type", alternate = {"packetType"})
    private String packetType;

    @SerializedName(value = "source_node_id", alternate = {"sourceNodeId"})
    private String sourceNodeId;

    @SerializedName(value = "destination_node_id", alternate = {"destinationNodeId"})
    private String destinationNodeId;

    @SerializedName(value = "sender_hop_id", alternate = {"senderHopId"})
    private String senderHopId;

    @SerializedName("ttl")
    private int ttl;

    @SerializedName(value = "hop_count", alternate = {"hopCount"})
    private int hopCount;

    @SerializedName("timestamp")
    private long timestamp;

    @SerializedName("payload")
    private Payload payload;

    @SerializedName(value = "route_history", alternate = {"routeHistory"})
    private List<String> routeHistory;

    @SerializedName("checksum")
    private String checksum;

    // ===== INNER CLASS: PAYLOAD =====

    /**
     * Nội dung nghiệp vụ của gói tin.
     */
    public static class Payload {

        @SerializedName(value = "sender_name", alternate = {"senderName"})
        private String senderName;

        @SerializedName(value = "alert_type", alternate = {"alertType"})
        private String alertType;

        @SerializedName("message")
        private String message;

        @SerializedName(value = "victim_count", alternate = {"victimCount"})
        private int victimCount;

        @SerializedName("severity")
        private String severity;

        @SerializedName("location")
        private Location location;

        @SerializedName(value = "ack_for_packet_id", alternate = {"ackForPacketId"})
        private String ackForPacketId;

        @SerializedName(value = "discovery_info", alternate = {"discoveryInfo"})
        private JsonElement discoveryInfo;

        public Payload() {}

        // --- Getters & Setters ---

        public String getSenderName() { return senderName; }
        public void setSenderName(String senderName) { this.senderName = senderName; }

        public String getAlertType() { return alertType; }
        public void setAlertType(String alertType) { this.alertType = alertType; }

        public String getMessage() { return message; }
        public void setMessage(String message) { this.message = message; }

        public int getVictimCount() { return victimCount; }
        public void setVictimCount(int victimCount) { this.victimCount = victimCount; }

        public String getSeverity() { return severity; }
        public void setSeverity(String severity) { this.severity = severity; }

        public Location getLocation() { return location; }
        public void setLocation(Location location) { this.location = location; }

        public String getAckForPacketId() { return ackForPacketId; }
        public void setAckForPacketId(String ackForPacketId) { this.ackForPacketId = ackForPacketId; }

        public JsonElement getDiscoveryInfo() { return discoveryInfo; }
        public void setDiscoveryInfo(JsonElement discoveryInfo) { this.discoveryInfo = discoveryInfo; }
        public void setDiscoveryInfo(Object discoveryInfo) {
            if (discoveryInfo == null) {
                this.discoveryInfo = null;
            } else if (discoveryInfo instanceof JsonElement) {
                this.discoveryInfo = (JsonElement) discoveryInfo;
            } else {
                this.discoveryInfo = ChecksumUtil.toJsonElement(discoveryInfo);
            }
        }

        @Override
        public String toString() {
            return "Payload{"
                    + "senderName='" + senderName + '\''
                    + ", alertType='" + alertType + '\''
                    + ", severity='" + severity + '\''
                    + ", victimCount=" + victimCount
                    + ", ackForPacketId='" + ackForPacketId + '\''
                    + ", location=" + location
                    + '}';
        }
    }

    // ===== INNER CLASS: LOCATION =====

    /**
     * Tọa độ địa lý GPS của sự cố/nạn nhân.
     */
    public static class Location {

        @SerializedName(value = "latitude", alternate = {"lat"})
        private double latitude;

        @SerializedName(value = "longitude", alternate = {"lon", "lng"})
        private double longitude;

        @SerializedName(value = "altitude", alternate = {"alt"})
        private double altitude = 0.0;

        @SerializedName(value = "accuracy", alternate = {"acc"})
        private double accuracy = 0.0;

        public Location() {}

        public Location(double latitude, double longitude) {
            this(latitude, longitude, 0.0, 0.0);
        }

        public Location(double latitude, double longitude, double altitude, double accuracy) {
            this.latitude  = latitude;
            this.longitude = longitude;
            this.altitude  = altitude;
            this.accuracy  = accuracy;
        }

        public double getLatitude()  { return latitude; }
        public void setLatitude(double latitude) { this.latitude = latitude; }

        public double getLongitude() { return longitude; }
        public void setLongitude(double longitude) { this.longitude = longitude; }

        public double getAltitude()  { return altitude; }
        public void setAltitude(double altitude) { this.altitude = altitude; }

        public double getAccuracy()  { return accuracy; }
        public void setAccuracy(double accuracy) { this.accuracy = accuracy; }

        @Override
        public String toString() {
            return "Location{lat=" + latitude + ", lon=" + longitude + ", alt=" + altitude + ", acc=" + accuracy + '}';
        }
    }

    // ===== CONSTRUCTORS =====

    public MeshPacket() {
        this.protocolVersion = PROTOCOL_VERSION_1;
        this.routeHistory = new ArrayList<>();
    }

    public MeshPacket(String packetType, String sourceNodeId, String destinationNodeId) {
        this.packetId          = UUID.randomUUID().toString();
        this.protocolVersion   = PROTOCOL_VERSION_1;
        this.packetType        = packetType;
        this.sourceNodeId      = sourceNodeId;
        this.destinationNodeId = destinationNodeId;
        this.senderHopId       = sourceNodeId;
        this.ttl               = DEFAULT_TTL;
        this.hopCount          = 0;
        this.timestamp         = System.currentTimeMillis();
        this.routeHistory      = new ArrayList<>();
    }

    // ===== CHECKSUM METHODS =====

    /**
     * Tính toán và gán checksum SHA-256 canonical cho toàn bộ gói tin.
     */
    public void computeAndSetChecksum() {
        this.checksum = ChecksumUtil.computePacketChecksum(this);
    }

    public void computeAndSetChecksum(Gson gson) {
        computeAndSetChecksum();
    }

    /**
     * Xác minh tính toàn vẹn của gói tin bằng cách tính lại checksum canonical.
     */
    public boolean verifyChecksum() {
        return ChecksumUtil.verifyPacketChecksum(this);
    }

    public boolean verifyChecksum(Gson gson) {
        return verifyChecksum();
    }

    // ===== SERIALIZATION METHODS =====

    public String toJson() {
        return toJson(new Gson());
    }

    public String toJson(Gson gson) {
        return gson.toJson(this);
    }

    public static MeshPacket fromJson(String json, Gson gson) {
        if (json == null || json.trim().isEmpty()) {
            return null;
        }
        try {
            MeshPacket packet = gson.fromJson(json, MeshPacket.class);
            if (packet != null) {
                packet.normalize();
            }
            return packet;
        } catch (Exception e) {
            System.err.println("[ERROR] MeshPacket.fromJson: không parse được JSON — " + e.getMessage());
            return null;
        }
    }

    /**
     * Chuẩn hóa gói tin sang canonical v1 (chuyển đổi alias cũ nếu có).
     */
    public void normalize() {
        if (TYPE_SOS_DATA.equals(this.packetType)) {
            this.packetType = TYPE_SOS_BROADCAST;
        } else if (TYPE_DISPATCH_CMD.equals(this.packetType)) {
            this.packetType = TYPE_DISPATCH_COMMAND;
        }
        if (this.protocolVersion == null || this.protocolVersion.trim().isEmpty()) {
            this.protocolVersion = PROTOCOL_VERSION_1;
        }
        if (this.routeHistory == null) {
            this.routeHistory = new ArrayList<>();
        }
    }

    // ===== GETTERS & SETTERS =====

    public String getPacketId() { return packetId; }
    public void setPacketId(String packetId) { this.packetId = packetId; }

    public String getProtocolVersion() { return protocolVersion; }
    public void setProtocolVersion(String protocolVersion) { this.protocolVersion = protocolVersion; }

    public String getPacketType() { return packetType; }
    public void setPacketType(String packetType) { this.packetType = packetType; }

    public String getSourceNodeId() { return sourceNodeId; }
    public void setSourceNodeId(String sourceNodeId) { this.sourceNodeId = sourceNodeId; }

    public String getDestinationNodeId() { return destinationNodeId; }
    public void setDestinationNodeId(String destinationNodeId) { this.destinationNodeId = destinationNodeId; }

    public String getSenderHopId() { return senderHopId; }
    public void setSenderHopId(String senderHopId) { this.senderHopId = senderHopId; }

    public int getTtl() { return ttl; }
    public void setTtl(int ttl) { this.ttl = ttl; }

    public int getHopCount() { return hopCount; }
    public void setHopCount(int hopCount) { this.hopCount = hopCount; }

    public long getTimestamp() { return timestamp; }
    public void setTimestamp(long timestamp) { this.timestamp = timestamp; }

    public Payload getPayload() { return payload; }
    public void setPayload(Payload payload) { this.payload = payload; }

    public List<String> getRouteHistory() { return routeHistory; }
    public void setRouteHistory(List<String> routeHistory) { this.routeHistory = routeHistory; }

    public String getChecksum() { return checksum; }
    public void setChecksum(String checksum) { this.checksum = checksum; }

    // ===== UTILITY METHODS =====

    public void addToRouteHistory(String nodeId) {
        if (this.routeHistory == null) {
            this.routeHistory = new ArrayList<>();
        }
        this.routeHistory.add(nodeId);
    }

    @Override
    public String toString() {
        return "MeshPacket{"
                + "id='" + (packetId != null && packetId.length() > 8 ? packetId.substring(0, 8) + "..." : packetId) + '\''
                + ", v='" + protocolVersion + '\''
                + ", type='" + packetType + '\''
                + ", from='" + sourceNodeId + '\''
                + ", to='" + destinationNodeId + '\''
                + ", ttl=" + ttl
                + ", hops=" + hopCount
                + ", route=" + routeHistory
                + '}';
    }
}

File: android-app/app/src/main/java/com/rescue/mesh/android/protocol/MeshPacket.kt
package com.rescue.mesh.android.protocol

import com.google.gson.Gson
import com.google.gson.JsonElement
import java.nio.charset.StandardCharsets
import com.google.gson.annotations.SerializedName
import java.util.UUID

/**
 * Protocol v1 packet shared with the Java desktop application.
 *
 * Wire names deliberately use snake_case and the checksum is calculated by
 * [ChecksumUtil] over the canonical representation, not over Gson's field
 * order.  The model is nullable at the wire boundary so malformed input can
 * be rejected by [PacketValidator] instead of crashing the socket service.
 */
data class MeshPacket(
    @SerializedName(value = "packet_id", alternate = ["packetId"])
    var packetId: String? = null,
    @SerializedName(value = "protocol_version", alternate = ["protocolVersion"])
    var protocolVersion: String? = PROTOCOL_VERSION_1,
    @SerializedName(value = "packet_type", alternate = ["packetType"])
    var packetType: String? = null,
    @SerializedName(value = "source_node_id", alternate = ["sourceNodeId"])
    var sourceNodeId: String? = null,
    @SerializedName(value = "destination_node_id", alternate = ["destinationNodeId"])
    var destinationNodeId: String? = null,
    @SerializedName(value = "sender_hop_id", alternate = ["senderHopId"])
    var senderHopId: String? = null,
    var ttl: Int = 0,
    @SerializedName(value = "hop_count", alternate = ["hopCount"])
    var hopCount: Int = 0,
    var timestamp: Long = 0L,
    var payload: Payload? = null,
    @SerializedName(value = "route_history", alternate = ["routeHistory"])
    var routeHistory: MutableList<String>? = mutableListOf(),
    var checksum: String? = null
) {
    fun normalize(): MeshPacket {
        packetType = when (packetType) {
            TYPE_SOS_DATA -> TYPE_SOS_BROADCAST
            TYPE_DISPATCH_CMD -> TYPE_DISPATCH_COMMAND
            else -> packetType
        }
        if (protocolVersion.isNullOrBlank()) protocolVersion = PROTOCOL_VERSION_1
        if (routeHistory == null) routeHistory = mutableListOf()
        return this
    }

    fun toJson(): String = GSON.toJson(this)

    fun computeAndSetChecksum(): String {
        checksum = ChecksumUtil.computePacketChecksum(this)
        return checksum!!
    }

    fun verifyChecksum(): Boolean = ChecksumUtil.verifyPacketChecksum(this)

    fun copyForForwarding(nodeId: String): MeshPacket {
        require(nodeId.isNotBlank()) { "forwarding node id must not be blank" }
        val forwarded = fromJson(toJson()) ?: error("cannot copy packet")
        forwarded.ttl -= 1
        forwarded.hopCount += 1
        val history = forwarded.routeHistory ?: mutableListOf()
        history += nodeId
        forwarded.routeHistory = history
        forwarded.senderHopId = nodeId
        forwarded.computeAndSetChecksum()
        return forwarded
    }

    companion object {
        const val PROTOCOL_VERSION_1 = "1.0"
        const val TYPE_SOS_BROADCAST = "SOS_BROADCAST"
        const val TYPE_DISPATCH_COMMAND = "DISPATCH_COMMAND"
        const val TYPE_ROUTE_DISCOVERY = "ROUTE_DISCOVERY"
        const val TYPE_ACK = "ACK"
        const val TYPE_HEARTBEAT = "HEARTBEAT"
        @Deprecated("Use TYPE_SOS_BROADCAST") const val TYPE_SOS_DATA = "SOS_DATA"
        @Deprecated("Use TYPE_DISPATCH_COMMAND") const val TYPE_DISPATCH_CMD = "DISPATCH_CMD"

        const val NODE_BASE_STATION = "BASE_STATION"
        const val NODE_BROADCAST = "BROADCAST"
        const val SEVERITY_CRITICAL = "CRITICAL"
        const val SEVERITY_HIGH = "HIGH"
        const val SEVERITY_MEDIUM = "MEDIUM"
        const val ALERT_MEDICAL = "MEDICAL"
        const val ALERT_FLOOD = "FLOOD_TRAPPED"
        const val ALERT_LANDSLIDE = "LANDSLIDE"
        const val DEFAULT_TTL = 5

        private val GSON = Gson()

        fun newSos(
            sourceNodeId: String,
            senderName: String,
            alertType: String,
            message: String,
            victimCount: Int,
            severity: String,
            location: Location
        ): MeshPacket = MeshPacket(
            packetId = UUID.randomUUID().toString(),
            protocolVersion = PROTOCOL_VERSION_1,
            packetType = TYPE_SOS_BROADCAST,
            sourceNodeId = sourceNodeId,
            destinationNodeId = NODE_BASE_STATION,
            senderHopId = sourceNodeId,
            ttl = DEFAULT_TTL,
            hopCount = 0,
            timestamp = System.currentTimeMillis(),
            payload = Payload(senderName, alertType, message, victimCount, severity, location),
            routeHistory = mutableListOf()
        ).also { it.computeAndSetChecksum() }

        fun newAck(myNodeId: String, originalPacketId: String, destinationNodeId: String): MeshPacket =
            MeshPacket(
                // ACK identity is stable for a (receiver, original packet)
                // pair.  A duplicate SOS can therefore reuse the durable ACK
                // outbox row instead of creating an unbounded stream of ACKs.
                packetId = UUID.nameUUIDFromBytes(
                    "ACK|$myNodeId|$originalPacketId".toByteArray(StandardCharsets.UTF_8)
                ).toString(),
                protocolVersion = PROTOCOL_VERSION_1,
                packetType = TYPE_ACK,
                sourceNodeId = myNodeId,
                destinationNodeId = destinationNodeId,
                senderHopId = myNodeId,
                ttl = DEFAULT_TTL,
                hopCount = 0,
                timestamp = System.currentTimeMillis(),
                payload = Payload(
                    senderName = myNodeId,
                    alertType = TYPE_ACK,
                    message = "ACK for packet: $originalPacketId",
                    victimCount = 0,
                    severity = SEVERITY_MEDIUM,
                    location = Location(),
                    ackForPacketId = originalPacketId
                ),
                routeHistory = mutableListOf()
            ).also { it.computeAndSetChecksum() }

        fun newDispatch(
            destinationNodeId: String,
            message: String,
            severity: String = SEVERITY_CRITICAL
        ): MeshPacket = MeshPacket(
            packetId = UUID.randomUUID().toString(),
            protocolVersion = PROTOCOL_VERSION_1,
            packetType = TYPE_DISPATCH_COMMAND,
            sourceNodeId = NODE_BASE_STATION,
            destinationNodeId = destinationNodeId,
            senderHopId = NODE_BASE_STATION,
            ttl = DEFAULT_TTL,
            hopCount = 0,
            timestamp = System.currentTimeMillis(),
            payload = Payload(
                senderName = "BASE_STATION_COMMANDER",
                alertType = "DISPATCH",
                message = message,
                victimCount = 0,
                severity = severity,
                location = Location()
            ),
            routeHistory = mutableListOf()
        ).also { it.computeAndSetChecksum() }

        fun fromJson(json: String): MeshPacket? = runCatching {
            if (json.isBlank()) null else GSON.fromJson(json, MeshPacket::class.java)?.normalize()
        }.getOrNull()

        fun fromJsonElement(element: JsonElement): MeshPacket? = runCatching {
            GSON.fromJson(element, MeshPacket::class.java)?.normalize()
        }.getOrNull()
    }
}

data class Payload(
    @SerializedName(value = "sender_name", alternate = ["senderName"])
    var senderName: String? = null,
    @SerializedName(value = "alert_type", alternate = ["alertType"])
    var alertType: String? = null,
    var message: String? = null,
    @SerializedName(value = "victim_count", alternate = ["victimCount"])
    var victimCount: Int = 0,
    var severity: String? = null,
    var location: Location? = null,
    @SerializedName(value = "ack_for_packet_id", alternate = ["ackForPacketId"])
    var ackForPacketId: String? = null,
    @SerializedName(value = "discovery_info", alternate = ["discoveryInfo"])
    var discoveryInfo: JsonElement? = null
)

data class Location(
    var latitude: Double = 0.0,
    var longitude: Double = 0.0,
    var altitude: Double = 0.0,
    var accuracy: Double = 0.0
)

enum class PacketType(val wireName: String) {
    SOS_BROADCAST(MeshPacket.TYPE_SOS_BROADCAST),
    DISPATCH_COMMAND(MeshPacket.TYPE_DISPATCH_COMMAND),
    ROUTE_DISCOVERY(MeshPacket.TYPE_ROUTE_DISCOVERY),
    ACK(MeshPacket.TYPE_ACK),
    HEARTBEAT(MeshPacket.TYPE_HEARTBEAT);

    companion object {
        fun fromWire(value: String?): PacketType? = entries.firstOrNull { it.wireName == value }
    }
}
